### PR TITLE
[FLINK-25719][python] Support General Python UDF in Thread Mode

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change

*This pull request will support General Python UDF in Thread Mode*

## Brief change log

  - *Refactor dependency managers*
  - *Support General Python UDF in Thread Mode*

## Verifying this change

This change added tests and can be verified as follows:

  - *`EmbeddedMultiThreadDependencyTests` and `EmbeddedSubInterpreterDependencyTests` in `test_dependency.py`*
  - *`PyFlinkEmbeddedMultiThreadTests` and `PyFlinkEmbeddedSubInterpreterTests` in `test_udf.py`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/AbstractEmbeddedPythonFunctionOperator.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+            if (dirAndNums.f0 == null) {
+                // get current directory.
+                interpreter.exec("import os;cwd = os.getcwd();");
+                dirAndNums.f0 = interpreter.get("cwd", String.class);
+                String workingDirectory = env.get(PYTHON_WORKING_DIR);
+                // set working directory
+                interpreter.exec(String.format("import os;os.chdir('%s')", workingDirectory));
+            }
+
+            lock.unlock();
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/AbstractEmbeddedPythonFunctionOperator.java:104]: move to finally clause

File: flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/AbstractEmbeddedPythonFunctionOperator.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            JobID jobId = getRuntimeContext().getJobId();
+            if (workingDirectories.containsKey(jobId)) {
+                lock.lockInterruptibly();
+                Tuple2<String, Integer> dirAndNums = workingDirectories.get(jobId);
+                dirAndNums.f1 -= 1;
+                if (dirAndNums.f1 == 0) {
+                    // change to previous working directory.
+                    interpreter.exec(String.format("import os;os.chdir('%s')", dirAndNums.f0));
+                }
+                lock.unlock();
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/AbstractEmbeddedPythonFunctionOperator.java:130]: move to finally

File: flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/AbstractEmbeddedPythonFunctionOperator.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+                        pythonConfig, getRuntimeContext().getDistributedCache());
+        return new EmbeddedPythonEnvironmentManager(
+                dependencyInfo,
+                getContainingTask().getEnvironment().getTaskManagerInfo().getTmpDirectories(),
+                new HashMap<>(System.getenv()),
+                getRuntimeContext().getJobId());
+    }
+
+    /** Creates the operation in multi-thread mode. */
+    public abstract void createOperationInMultiThread();
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/AbstractEmbeddedPythonFunctionOperator.java:158]: What about rename it to openPythonInterpreter and merge createOperationInMultiThread and createOperationInMultiInterpreters into one method

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/scalar/EmbeddedPythonScalarFunctionOperator.java [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+            rowDataWrapper.collect(reuseJoinedRow);
+        } else {
+            rowDataWrapper.collect(reuseResultRowData);
+        }
+    }
+
+    @Override
+    protected void invokeFinishBundle() throws Exception {
+        // TODO: Support batches invoking.
+        if (elementCount > 0) {
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/scalar/EmbeddedPythonScalarFunctionOperator.java:250]: could be removed for now?

Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/scalar/EmbeddedPythonScalarFunctionOperator.java:250]: could be removed for now?

File: flink-python/src/main/java/org/apache/flink/python/PythonOptions.java [hunk 5]
```diff
@@ -214,4 +214,25 @@
                                     + "in each batch when iterating a Python MapState. Note that this is an experimental flag "
                                     + "and might not be available "
                                     + "in future releases.");
+
+    /** Specify the python runtime execution mode. */
+    @Experimental
+    public static final ConfigOption<String> PYTHON_EXECUTION_MODE =
+            ConfigOptions.key("python.execution-mode")
+                    .defaultValue("process")
```
[the remark this answers is not in the document]
Comment (HuangXingBo) [reply flink-python/src/main/java/org/apache/flink/python/PythonOptions.java:222]: Make sense

**Review by dianfu — APPROVED** [review]

**Pull request closed without merging**